### PR TITLE
Popup ready signal to service worker

### PR DESCRIPTION
**Pull request by JasonMHasperhoven**
Fixes https://github.com/penumbra-zone/web/issues/285

### CODE REVIEW

File: apps/extension/src/message/popup.ts [hunk 1]
```diff
@@ -11,9 +11,10 @@ import { OriginRecord } from '../storage/types';
 export enum PopupType {
   TxApproval = 'TxApproval',
   OriginApproval = 'OriginApproval',
+  Ready = 'PopupReady',
```
Comment (grod220) [inline apps/extension/src/message/popup.ts:14]: suggestion: should we rename this to say `Ready = "PopupReadyToAcceptMessages"`?

File: apps/extension/src/popup.ts [hunk 2]
```diff
@@ -18,9 +24,10 @@ const isChromeResponderDroppedError = (
 export const popup = async <M extends PopupMessage>(
   req: PopupRequest<M>,
 ): Promise<M['response']> => {
-  await spawnPopup(req.type);
-  // We have to wait for React to bootup, navigate to the page, and render the components
-  await new Promise(resolve => setTimeout(resolve, 800));
+  const popupId = crypto.randomUUID();
+  await spawnPopup(req.type, popupId);
+  await popupReady(popupId);
```
Comment (grod220) [inline apps/extension/src/popup.ts:29]: suggestion: we should add a code comment here that informs devs that this is necessary given it takes a bit of time for the popup to be ready to accept messages.

**Review by grod220 — APPROVED** [review]
Can you confirm this solves the problem with https://github.com/prax-wallet/web/issues/130? On the testnet whale, you can test this by trying to do an `EndAll` on the auctions page.

Approving, but think this also requires an approval from @turbocrime as they had a vision for how this was going to work from way back.

**Review by turbocrime — CHANGES_REQUESTED** [review]
blocking due to issues described by inline comments.

i think part of these issues are due to the lack of documentation for `InternalMessage` and associated types. it is related to some of the earliest messaging work on the extension, so maybe @grod220 has some input. i will take the opportunity to explain some of the intentions as i understand them in a following comment.

**Comment by turbocrime** [comment]
sorry if this is excessively thorough or verbose. you mentioned a lower
familiarity with typescript, the problems here are decently subtle, and i'm
offboarding so i'm taking this opportunity to deliver as much detail as
possible.

### `InternalMessage`

`InternalMessage` is used to type specific, scoped document-to-document
messaging within prax. it may be used to describe exactly two message types
involved in a simple request-response dialogue, which can be retrieved by the
utility types `InternalRequest` and `InternalResponse`.

`InternalMessage`'s first parameter is type name string, second parameter
request shape, and third parameter response shape. These parameters are
independent and any of these type parameters may be type unions. union
parameters are used to aggregate purpose-specific messages, and are shared by
message handlers and control utilities.

The actual contents are often derived from user input, and importantly, an
`InternalMessage` handler will encounter messages of unknown shape for unknown
destinations, so the presence of a type field string does not actually validate
the remaining message structure - the request type parameter suggests the shape
that should be validated.

response types are less intensively validated because chrome allows the caller
to await a 'sendMessage' call to receive a specific, call-scoped response. so in
handlers the response type utilities are used to create appropriate responses,
and in message-sending utilities the response types are used to type return
values, but these are not guarded.

Some basic structure is enforced by `InternalMessage` and the design is intended
to accommodate error transmission, but message-sending utilities are designed to
conceal those details and simply accept bare parameters, return bare values, and
rethrow deserialized errors.

So: A type string is just a label, and doesn't validate type. A handler will
process requests carefully. A caller can accept responses confidently.

### message types are grouped by purpose

There are a few sets of messages grouped by purpose, indicated by a type union
of those messages. Generally you can expect that a single handler will process
all request members of the union, from unknown input, disambiguated through that
purpose-specific message type union.

if the messages relate to a control utility, there may be a control utility
typed with param/return types matching the inner request/response contents.

### `PopupMessage`

The popup specifically uses type unions `PopupRequest` and `PopupResponse` which
may be optionally narrowed. they are synonymous with `InternalRequest` and
`InternalResponse` restricted to and defaulting to the union of `PopupMessage`.

The main extension popup control utility is simply called `popup`, located at
`src/popup.ts`, and it uses these types to describe its valid input, and the
input type corresponds to a return type derived from each request's appropriate
response. Adding members of `InternalMessage` to the type union `PopupMessage`
will expand the valid inputs of the `popup` method input, and instruct
typescript to expect the corresponding return.

### `PopupType`

`PopupType` enumerates the available types of popup provided by the popup
control utility. at the moment this corresponds nicely with single decision
methods requiring input and output, so it is used both as a type string for
`PopupMessage` and a parameter for the `popup` control utility.

for `popup`, `PopupType` controls:

- the page route navigated by the launched window
- the request type indexed to type the parameter
- the response type indexed to type the return

for popup message type guards, `PopupType` controls the contents of the union
that a type guard must discriminate in order to identify a message.

for message handlers, `PopupType` affects

- what kind of message is considered relevant
- the union which the handler must discriminate after detecting a relevant message
- what kind of input should be applied to the response callback

## The Problems

so, adding `PopupType.Ready` to the popup type enum and adding the type
`InternalMessage<PopupType.Ready, null, { popupId: string; }>` to the
`PopupMessage` union means that the popup utility now understands it may
identify a route for 'Ready' to ask the user to take some action regarding
`null`. the return type does not include `null` so the user will never close the
window without responding, and result of popup will reliably be a response
object with a `popupId`.

```ts
import { popup } from './popup'; // extension src/popup.ts
const readyResponse: { popupId: string } = await popup<Ready>(null);
```

thankfully, `PopupType.Ready` does not correspond to any extension route, so the
controller just throws.

additionally, the expanded type union and deletion of discriminating checks in
the type guards

1. allows the popup message handler to accept more inputs, for which no handler code is added
2. claims to the popup message handler that possibly undefined input fields are definitely not undefined, for which no behavior is certain

so actual message entry into the handler is still affected, and the results of
that are uncertain. this is possibly dangerous.

## The Suggestions

it makes a lot of sense to simply add a new type string to what looks like a
collection of message types, and none of this was documented or explained.

### suggestion: Don't use `InternalMessage` at all.

You're providing a UUID to the popup via its URI hash. That specific popup can
uniquely identify itself, guaranteed, for this specific purpose, by transmitting
a message containing that uuid alone. No type guard, no object comparison.

simplifying `popupReady`

```ts
const popupReady = (id: string): Promise<void> =>
  new Promise((resolve, reject) => {
    AbortSignal.timeout(POPUP_READY_TIMEOUT).onabort = reject;

    const idListen = (res: unknown, _: chrome.runtime.MessageSender, respond: () => void) => {
      if (res === id) {
        resolve();
        chrome.runtime.onMessage.removeListener(idListen);
        respond();
      }
    };

    chrome.runtime.onMessage.addListener(idListen);
  });
```

And this requires no message types changes or handler refactors. In the popup,
you'd have a similar pattern to now.

It seems intuitive to provide the 'input' as the response instead of nothing,
but then we'd have to refactor the existing types handlers etc, so i think this
is better.

### alternative suggestion: Use `InternalMessage` and define the `{ popupId }`

#payload as the request, not the response.

the request type is intended for parameter input and use in handlers and type
guards, and should exclude null. the response type is intended for use in
response callbacks and returns, and needs no type guards because it's obtained
from typed functions or deserialized from a targeted message.

but you've defined `Ready` as null-request with typed response, even though it's
being emitted by `sendMessage` to initiate a request cycle. the direction of the
message - worker to popup, or popup to worker - is not relevant here, but maybe
it was misleading.

flip the definition:

```ts
export const Ready = InternalMessage<PopupType.Ready, { popupId: string }, null>;
```

after that there is much more to this, but now the type is used more
appropriately. as `Ready` still doesn't make sense as a popup _control_ message,
there's still the problem with the concept of `Ready`-type messages as an input
type and affecting the broader popup type logic, but at least it's not null.

### todo: throwIfAlreadyOpen

this is pretty critical, i think the explanation above suffices

### todo: use hash router instead of query params

the hash router is already used to path all the popup routes, and we should use
one utility only to manipulate and read parameters.

File: apps/extension/src/message/popup.ts [hunk 3]
```diff
@@ -42,8 +51,19 @@ export const isPopupRequest = (req: unknown): req is PopupRequest =>
...
+  typeof res.type === 'string' &&
+  res.type in PopupType;
+
 export const isOriginApprovalRequest = (req: unknown): req is InternalRequest<OriginApproval> =>
-  isPopupRequest(req) && req.type === PopupType.OriginApproval && 'origin' in req.request;
+  isPopupRequest(req) && req.type === PopupType.OriginApproval;
 
 export const isTxApprovalRequest = (req: unknown): req is InternalRequest<TxApproval> =>
-  isPopupRequest(req) && req.type === PopupType.TxApproval && 'authorizeRequest' in req.request;
+  isPopupRequest(req) && req.type === PopupType.TxApproval;
```
[the remark this answers is not in the document]
Comment (grod220) [reply apps/extension/src/message/popup.ts:66]: The extra field checks feel unnecessary here. By asserting the presence of specific fields, we are saying that say `PopupType.OriginApproval` could be the request type, but there could still be a request body that does not fit. Given our defined types, that's not possible.

File: apps/extension/src/routes/popup/popup-layout.tsx [hunk 4]
```diff
@@ -11,8 +12,12 @@ import { Outlet } from 'react-router-dom';
  * routes here in `PopupLayout`, and using a different root class name for each,
  * then removing the hard-coded width from `globals.css`.
  */
-export const PopupLayout = () => (
-  <div className='flex grow flex-col bg-card-radial'>
-    <Outlet />
-  </div>
-);
+export const PopupLayout = () => {
+  usePopupReady();
```
[the remark this answers is not in the document]
Comment (JasonMHasperhoven) [reply apps/extension/src/routes/popup/popup-layout.tsx:16]: I believe that placing it here keeps the code simpler and makes adding any new route in the future work out of the box, so also more convenient.

**Comment by JasonMHasperhoven** [comment]
> Can you confirm this solves the problem with #130? On the testnet whale, you can test this by trying to do an `EndAll` on the auctions page.
> 
> Approving, but think this also requires an approval from @turbocrime as they had a vision for how this was going to work from way back.

Seems to work though it takes time @grod220

**Comment by JasonMHasperhoven** [comment]
Thank you for the extensive write-up @turbocrime. I understand now that this message should not be of the InternalMessage type. I've refactored it according to your suggestions, please re-review.

**Comment by turbocrime** [comment]
it looks like `throwIfAlreadyOpen` is not updated to consider the path containing uuid will always be unique

File: apps/extension/src/hooks/popup-ready.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  const popupId = searchParams.get('popupId');
+
+  useEffect(() => {
+    if (popupId && (isReady === undefined || isReady) && !sentMessagesRef.current.has(popupId)) {
+      sentMessagesRef.current.add(popupId);
+
+      void chrome.runtime.sendMessage(popupId);
+    }
+  }, [popupId, isReady]);
+};
```
[the remark this answers is not in the document]
Comment (turbocrime) [reply apps/extension/src/hooks/popup-ready.ts:19]:
```suggestion
import { useEffect, useRef } from 'react';
import { useSearchParams } from 'react-router-dom';

// signals that react is ready (mounted) to service worker
export const usePopupReady = () => {
  const sentReady = useRef<string>();
  const [searchParams] = useSearchParams();
  const popupId = searchParams.get('popupId');

  useEffect(() => {
    if (!sentReady.current && popupId) {
      sentReady.current ??= popupId;
      void chrome.runtime.sendMessage(sentReady.current);
    }
  }, [popupId, searchParams]);
};
```

**Review by turbocrime — APPROVED** [review]
approving, but please correct use or implementation of `throwIfAlreadyOpen`

File: apps/extension/src/hooks/popup-ready.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JasonMHasperhoven) [reply apps/extension/src/hooks/popup-ready.ts:19]: Actually we can remove the function arg, the initial idea (before I placed it into the layout) was that you would be able to control the ready state from a parent component, but this is no longer needed and overcomplicates things.

**Comment by JasonMHasperhoven** [comment]
> it looks like `throwIfAlreadyOpen` is not updated to consider the path containing uuid will always be unique

This is no longer needed, because the `?popup` query is moved to the `popUrl.hash` var as this makes it available for the react-router (useSearchParams).

**Comment by grod220** [comment]
A subtle, but noticeable difference is that the extension popup content loads faster (at least for me). Nice job!

**Pull request merged**